### PR TITLE
Add coordination error types, validation helpers, and idempotent outcome wrapper

### PR DESCRIPTION
**Pull request by ahrav**
## Summary

Ports the coordination error model and validation functions from design artifacts into `gossip-contracts`. This establishes the type-safe error hierarchy and pure validation helpers that the coordinator backend will use for all shard lifecycle operations.

- **`error.rs`** (1,297 lines): `CoordError` shared enum (11 variants), 7 operation-specific error newtypes (`AcquireError`, `RenewError`, `CheckpointError`, `CompleteError`, `ParkError`, `SplitReplaceError`, `SplitResidualError`), `IdempotentOutcome<T>` wrapper, and result types
- **`validation.rs`** (715 lines): `validate_lease()`, `validate_cursor_update()`, `check_op_idempotency()` — pure, side-effect-free precondition checks
- **`mod.rs`**: declares modules and re-exports public types

## Design Decisions

| Decision | Rationale |
|----------|-----------|
| Operation-specific error newtypes over shared `CoordError` | Callers get precise matching — `CheckpointError` can't produce `AlreadyLeased`. Tradeoff: boilerplate `From` impls, but finite and mechanical. (D2.16) |
| Explicit rejection arms in `From` impls (no `_ =>`) | Adding a new `CoordError` variant forces a conscious routing decision in every operation type — compile-time exhaustiveness |
| `#[non_exhaustive]` on `CoordError` and `IdempotentOutcome` | Forward compatibility for downstream crates |
| `TenantMismatch` omits `actual` tenant | SEC-1: prevents cross-tenant enumeration |
| `AlreadyLeased` omits owner from `Display` | SEC-5: worker identity redaction |
| `validate_lease` check ordering: tenant → terminal → fence → expiry | Security-first: tenant mismatch never reveals shard state |
| Pure validation functions returning `Result<(), CoordError>` | Side-effect-free, composable; callers convert via `From<CoordError>` with `?` |

## Architecture

```mermaid
graph TD
    subgraph "Caller (coordinator backend)"
        OP[Operation: checkpoint / complete / park / split]
    end

    subgraph "validation.rs (pure functions)"
        VL[validate_lease]
        COI[check_op_idempotency]
        VCU[validate_cursor_update]
    end

    subgraph "error.rs"
        CE[CoordError — shared enum]
        OE["Operation Errors<br/>(CheckpointError, CompleteError, etc.)"]
        IO["IdempotentOutcome&lt;T&gt;"]
    end

    OP --> VL
    OP --> COI
    OP --> VCU

    VL -->|Err| CE
    COI -->|Err| CE
    VCU -->|Err| CE

    CE -->|"From&lt;CoordError&gt;"| OE
    OP -->|Ok| IO
```

## Test Plan

- [ ] **error.rs** (26 tests): `IdempotentOutcome` API, `From<CoordError>` exhaustiveness for all 6 operation types, security (no tenant/owner/hash leaks in Display), display determinism, `source()` chain propagation, `PartialEq` value-equality for `Box<[u8]>` fields
- [ ] **validation.rs** (18 tests + 1 proptest): `validate_lease` happy path + each error case + all 11 multi-condition priority combinations, `validate_cursor_update` forward/regression/bounds/boundary cases, `check_op_idempotency` new/replay/conflict/capacity, property-based test validating result consistency against cursor primitives
- [ ] `cargo fmt --all && cargo check && cargo clippy --all-targets --all-features -- -D warnings`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/gossip-contracts/src/coordination/validation.rs [hunk 1]
```diff
@@ -0,0 +1,715 @@
...
+    }
+
+    // 4. Lease expiry. If the fence epoch check passed, the record
+    // MUST have a lease holder — unwrap_or(ZERO) would mask a logic bug.
+    let deadline = record
+        .lease_deadline()
+        .expect("lease deadline must exist when fence epochs match");
+    if now >= deadline {
+        return Err(CoordError::LeaseExpired { deadline, now });
+    }
```
[the remark this answers is not in the document]
Comment (ahrav) [reply crates/gossip-contracts/src/coordination/validation.rs]: The code already handles this defensively — line 92 uses `let Some(deadline) = record.lease_deadline() else { return Err(StaleFence { ... }) }`, not `expect()`. There is no panic path here.

The existing test `validate_lease_no_lease_returns_stale_fence` (validation.rs:401) covers this exact scenario: active unleased record with `INITIAL` fence, caller presents matching `INITIAL` fence → returns `StaleFence`. Test passes.

File: crates/gossip-contracts/src/coordination/error.rs [hunk 2]
```diff
@@ -0,0 +1,1297 @@
...
+
+    /// Another worker currently holds a valid (non-expired) lease.
+    /// The caller must wait or try a different shard.
+    ///
+    /// **Security note (SEC-5):** `current_owner` exposes worker identity.
+    /// Redact this field before surfacing to external clients.
+    AlreadyLeased {
+        current_owner: WorkerId,
+        lease_deadline: LogicalTime,
+    },
```
[the remark this answers is not in the document]
Comment (ahrav) [reply crates/gossip-contracts/src/coordination/error.rs]: `AcquireError` does not use `#[derive(Debug)]` — it has a manual `impl fmt::Debug` (error.rs:288) that redacts `current_owner` with `"<redacted>"`. The derive line is `#[derive(Clone, PartialEq, Eq)]` only.

The test `already_leased_debug_no_owner_leak` (error.rs:1445) verifies no worker identity appears in `Debug` output.
